lib/hackmode/index.js

```diff
@@ -503,69 +581,227 @@ class N3hHackMode extends N3hMode {
+        // log.t(this.me() + '.gossipTrack', this._ipcDnaByAgent.size)
+        this._p2pSend(evt.peerAddress, {
+          type: 'gossipAllTracks',
+          dnaByAgent: Array.from(this._ipcDnaByAgent.keys())
+        })
         break
       case 'handlePublish':
-        this._handleP2pMessage({
+        // Handle 'publish' message sent from some other peer's _p2pSend()
+        this._handleP2pPublish({
           from: evt.fromPeerAddress,
           data: msgpack.decode(Buffer.from(evt.data, 'base64'))
         })
         break
+      // case 'handleRequest':
       default:
         throw new Error('unexpected event type: ' + evt.type)
     }
   }
 
   /**
-   * Received a message from the network.
+   * Received P2P message from other Peer
    * Might send some messages back and
    * transcribe received message into a local IPC message.
    * @private
    */
-  _handleP2pMessage (opt) {
-    // log.w('@@@@', opt.data.type, JSON.stringify(opt.data))
+  _handleP2pPublish (opt) {
+    // log.t(this.me() + ' << ' + this.nick(opt.from) + ' - ' + opt.data.type)
     switch (opt.data.type) {
-      case 'gossipHashHash':
-        // send back a gossipRequestLocHashes and a gossipHashHashResp
-        this._processGossipHashHash(opt.from, opt.data.gossipHashHash)
-        const gossipHashHash = this._fullGossipHashHash()
-        // log.t('gossip (resp) with', opt.from, JSON.stringify(gossipHashHash))
-        this._p2pSend(opt.from, {
-          type: 'gossipHashHashResp',
-          gossipHashHash
-        })
-        return
-      case 'gossipHashHashResp':
-        // send back a gossipRequestLocHashes
-        this._processGossipHashHash(opt.from, opt.data.gossipHashHash)
+      case 'dhtGossip':
+        // log.t(this.me() + ' bundle =', opt.data.bundle)
+        this._dht.post(DhtEvent.remoteGossipBundle(
+          opt.from, opt.data.bundle))
         return
-      case 'gossipRequestLocHashes':
-        // send back a gossipHashList
-        this._processRequestLocHashes(opt.from, opt.data.locList)
-        return
-      case 'gossipHashList':
-        // send back many getData
-        this._processGossipHashList(opt.from, opt.data.hashList)
-        return
-      case 'getData':
-        // send back a getDataResp
-        this._processGetData(opt.from, opt.data.dnaAddress, opt.data.hash)
+      case 'gossipNewTrack':
+        // Some peer is telling us its tracking a new DNA
+        log.t(this.me() + ' @@@@ ' + this.me() + ' new track', opt.data.agentId, opt.data.dnaAddress)
+        // get peer transport
+        if (!this._peerList.includes(opt.from)) {
+          log.w('received gossipNewTrack from unknown peer', this.nick(opt.from))
+          return
+        }
+        // Bookkeep agentId -> PeerId
+        this._agent2peer.set(opt.data.agentId, opt.from)
+        // Add peer to dht
+        const peerHoldEvent = DhtEvent.peerHoldRequest(
+          opt.from,
+          'wss://0.0.0.0',
```

Comment:
I think there's a disconnect here. `peerHoldRequest( peerAddress, peerTransport, ... )`

So p2p-backend-hackmode generates an id like `abc123` and creates

`peerHoldRequest( 'abc123', 'wss://yada', ... )`

now, at the agentid level, we need to associate an agent id like `HcS0yada` with the transport id

`peerHoldRequest( 'HcS0yada', 'abc123', ... )`

then we won't have to do this manual peer discovery gossip, it should just be synced to all nodes, and if you do an `await this._dht.fetchPeer('HcS0yada')` you'll get back the 'abc123' and be able to send a message to them.

Reply:
We can't do `peerHoldRequest( 'HcS0yada', 'abc123', ... )` because second argument ,transport id, has to be an URL.
And most of the time we want to retrieve a peerAddress from an agentId, not the other way around.